### PR TITLE
planner: fix wrong charset about union result of date type and int

### LINKED ISSUE #27167: Wrong charset about union result of date type and int
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

run master and 
```
CREATE TABLE `all_types` (
  `id` int(11) NOT NULL,
  `d_tinyint` tinyint(4) DEFAULT NULL,
  `d_smallint` smallint(6) DEFAULT NULL,
  `d_int` int(11) DEFAULT NULL,
  `d_bigint` bigint(20) DEFAULT NULL,
  `d_float` float DEFAULT NULL,
  `d_double` double DEFAULT NULL,
  `d_decimal` decimal(10,2) DEFAULT NULL,
  `d_bit` bit(10) DEFAULT NULL,
  `d_binary` binary(10) DEFAULT NULL,
  `d_date` date DEFAULT NULL,
  `d_datetime` datetime DEFAULT NULL,
  `d_timestamp` timestamp NULL DEFAULT NULL,
  `d_varchar` varchar(20) NULL default NULL,
  PRIMARY KEY (`id`)
);

insert into all_types values(0, 0, 1, 2, 3, 1.5, 2.2, 10.23, 12, 'xy', '2021-12-12', now(), now(), '123');
```

### 2. What did you expect to see? (Required)

```
mysql> select d_date from all_types union select d_int from all_types;
+------------+
| d_date     |
+------------+
| 2021-12-12 |
| 2          |
+------------+
2 rows in set (0.07 sec)

```
### 3. What did you see instead (Required)

```
mysql> select d_date from all_types union select d_int from all_types;
+------------------------+
| d_date                 |
+------------------------+
| 0x32                   |
| 0x323032312D31322D3132 |
+------------------------+
2 rows in set (16.10 sec)
```

### 4. What is your TiDB version? (Required)
Master branch

### PR DESCRIPTION
**Pull request by yuqi1129**
### What problem does this PR solve?

Issue Number: close #27167

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Check List 

Tests 

- [x] Integration test

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by yuqi1129** [comment]
This kind of resolution seems not very gracefully and a little tricky, can you give me some advice?

**Review by wshwsh12 — COMMENTED** [review]
I think meeting the following three conditions should use default charset/collation:
1. FieldType's EvalType is IsStringKind.
2. Arguments don't contain BinaryStr.
3. DeriveCollation is CharsetBin/CollationBin.

And ideas for this or this issue? @xiongjiwei @wjhuang2016

**Review by xiongjiwei — COMMENTED** [review]
I think it should work like this: 
1. once the `union result type` is confirmed, all the columns should be cast to that type.
2. use the casted type to infer the collation.

in this case, the `union result type` is `string` and the `date` and `int` column should be cast to `string` so that it will have the connection charset and collation, then we will get the correct collation.

**Comment by yuqi1129** [comment]
@wshwsh12  As far as i know, even though i use `cast expr` to calculate return type,  it still failed to handle that SQL
image
image

Can you help to find this situation ?

**Comment by xiongjiwei** [comment]
ok, the problem may be: in `buildLogicalUnionAll` function, it should not use `expression.BuildCastFunction4Union(b.ctx, srcCol, dstType)` directly, because this function cast the value to the `dstType` including the charset and collation, but at this time, this information of `union result type` is not determined. The cast here should only consider the **`type`**, see `WrapWithCastAsString` for the difference. You may find it the `src type` is not `string`, the cast result's charset and collation will use `CharacterSetConnection` and `CollationConnection`

**Comment by yuqi1129** [comment]
@xiongjiwei , if we ignore charset and collation in cast, that is if charset and collation of type a(return type of cast function) and b are empty or meaningless, how can we refer the final charset and collation from them.

**Comment by xiongjiwei** [comment]
> @xiongjiwei , if we ignore charset and collation in cast...

nope, we did not ignore the charset and collation in cast, what I said is we can not use the undetermined charset and collation.
let me explain it using this case: when union `date` and `int` type, the result type will be **`string`**, but at this time we do not know what charset and collation should be used so that we should cast `date` and `int` to `string` (which only cast the type but charset and collation, the charset information should be inferred automatically, see `WrapWithCastAsString` function), after the cast, we get 2 string type value with charset and collation, finally, we can infer the result charset and collation.

**Comment by yuqi1129** [comment]
@xiongjiwei I see. 
> we get 2 string type values with charset and collation, finally, we can infer the result charset and collation

I got stuck by this situation. As we have two string type with binary charset and binary collation, how can we derive the result collation and charset. 

In this case, i can't distinguish the case: `select varchar_type union varchar_type` with cases in this issue. because after cast, return type, collation, charset are all the same. Originally,  for SQL like `select varchar_type union varchar_type`, we need not to do any changes about charset and collation. 

Simple resolution can be just take vachar union varchar as a special case and force the result charset and collation are right, but some execution plan will change and CI failed.

**Comment by xiongjiwei** [comment]
a quick solution may looks like 
image
which `expression.WarpWithCast` is 
image
it is same with https://github.com/pingcap/tidb/blob/60065dfbee0a37dbad3380f6a389b970e581d64e/expression/builtin.go#L156-L173

it works well for `select d_date from all_types union select d_int from all_types;`. I did not check the other scenario.

**Comment by xiongjiwei** [comment]
> As we have two string type with binary charset and binary collation

we will not get 2 binary charset values. You may misunderstand something please take a look at the solution above.

**Comment by yuqi1129** [comment]
@xiongjiwei Indeed, This changes will change some SQL planner , you can just see my CI test. 

> which expression.WarpWithCast is

Yes, what i do is almost the same.
image

CI has been failed since this change.

**Comment by xiongjiwei** [comment]
> This changes will change some SQL planner 

I think it is expected, I debug the test and find the charset information is not set correctly and make a redundant `cast` function appeared.

**Comment by yuqi1129** [comment]
@xiongjiwei ， Hi, Can you take some time to review this PR ?

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -1384,8 +1385,67 @@ func (b *PlanBuilder) buildProjection4Union(ctx context.Context, u *LogicalUnion
...
+		return expression.WrapWithCastAsTime(ctx, col, types.NewFieldType(mysql.TypeDatetime))
+	case types.ETTimestamp:
+		return expression.WrapWithCastAsTime(ctx, col, types.NewFieldType(mysql.TypeTimestamp))
+	case types.ETDuration:
+		return expression.WrapWithCastAsDuration(ctx, col)
+	case types.ETJson:
+		return expression.WrapWithCastAsJSON(ctx, col)
+	default:
+		return col
+	}
```
Comment (xiongjiwei) [inline planner/core/logical_plan_builder.go:1443]: this part is the same with `newBaseBuiltinFuncWithTp`, maybe we can re-use the code.

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -1384,8 +1385,67 @@ func (b *PlanBuilder) buildProjection4Union(ctx context.Context, u *LogicalUnion
 	}
 	u.schema = expression.NewSchema(unionCols...)
 	u.names = names
-	// Process each child and add a projection above original child.
-	// So the schema of `UnionAll` can be the same with its children's.
+
+	allExprs := make([][]expression.Expression, 0)
```
Comment (xiongjiwei) [inline planner/core/logical_plan_builder.go:1389]: why not use https://github.com/pingcap/tidb/pull/27212#issuecomment-902726520, we collect all the expressions and do the cast if needed, I think they are same

**Comment by xiongjiwei** [comment]
btw, could you please also test this query `select d_date from all_types union select d_int collate `binary` from all_types;`

File: planner/core/integration_test.go [hunk 3]
```diff
@@ -3926,6 +3926,38 @@ func (s *testIntegrationSerialSuite) TestIssue25300(c *C) {
...
+  		d_date date DEFAULT NULL,
+  		d_datetime datetime DEFAULT NULL,
+  		d_timestamp timestamp NULL DEFAULT NULL,
+  		d_varchar varchar(20) NULL default NULL,
+  		PRIMARY KEY (id)
+		);`)
+
+	tk.MustExec(`insert into all_types values(0, 0, 1, 2, 3, 1.5, 2.2, 10.23, 12, 'xy', '2021-12-12', '2021-12-12 12:00:00', '2021-12-12 12:00:00', '123');`)
+
+	tk.MustQuery("select * from (select d_date c from all_types union select d_int c from all_types) t order by c;").Check(testkit.Rows("2", "2021-12-12"))
```
[the remark this answers is not in the document]
Comment (xiongjiwei) [reply planner/core/integration_test.go:3954]: do you have any idea? is it another bug?

[the remark this answers is not in the document]
Comment (yuqi1129) [reply planner/core/integration_test.go:3954]: @xiongjiwei No,  I have not find the cause yet

**Comment by xiongjiwei** [comment]
hi, I am refactoring the collation infer logic now, find that this issue can be fixed by the refactoring, but I am not sure about the others. I would ask for your review after my refactor is done.

**Comment by yuqi1129** [comment]
@xiongjiwei , OK, Thanks for your review about this issue, i will do some effort to find out the reason before you submit the PR

**Pull request closed without merging**